File: nengo_conductance_synapses/sim_cond_exp/sim_cond_exp_pipe.py
#!/usr/bin/env python3

#   This file is part of soft_cond_lif
#   (c) Andreas Stöckel 2017
#
#   soft_cond_lif is free software: you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation, either version 3 of the License, or
#   (at your option) any later version.
#
#   soft_cond_lif is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
#
#   You should have received a copy of the GNU General Public License
#   along with soft_cond_lif.  If not, see <http://www.gnu.org/licenses/>.

from copy import deepcopy
import json
import hashlib
import os
import subprocess
import numpy as np
import shelve
import threading
import time
import pexpect


def task(cm=1e-9,
         g_leak=5e-8,
         v_thresh=-50e-3,
         v_rest=-65e-3,
         v_reset=-65e-3,
         tau_refrac=2e-3,
         tau_syn_e=5e-3,
         w_syn_e=0.1e-6,
         e_rev_syn_e=0e-3,
         tau_syn_i=5e-3,
         w_syn_i=0.1e-6,
         e_rev_syn_i=-75e-3,
         rate_e=0.0,
         rate_i=0.0,
         repeat=10,
         spike_loss_e=0.0,
         spike_loss_i=0.0,
         T=10.0,
         seed=None):
    """
    Assembles a Task data structure that will be sent to the neuron simulator
    process. The given id is used to associate individual experiment results
    with the original task.
    """

    # Use a random seed if None is given as seed
    if seed is None:
        seed = np.random.randint(np.iinfo(np.int32).max)

    return {
        "neuron": {
            "cm": cm,
            "g_leak": g_leak,
            "v_thresh": v_thresh,
            "v_rest": v_rest,
            "v_reset": v_reset,
            "tau_refrac": tau_refrac
        },
        "syn_e": {
            "tau_syn": tau_syn_e,
            "w_syn": w_syn_e,
            "e_rev_syn": e_rev_syn_e
        },
        "syn_i": {
            "tau_syn": tau_syn_i,
            "w_syn": w_syn_i,
            "e_rev_syn": e_rev_syn_i
        },
        "input": {
            "id": "",
            "T": T,
            "rate_e": rate_e,
            "rate_i": rate_i,
            "repeat": repeat,
            "spike_loss_e": spike_loss_e,
            "spike_loss_i": spike_loss_i,
            "seed": seed
        }
    }


def translate_simulation_parameters(e_rev_e=4.33,
                                    e_rev_i=-0.33,
                                    w_syn_e=1.0,
                                    w_syn_i=1.0,
                                    gL=50,
                                    tau_syn_e=None,
                                    tau_syn_i=None,
                                    params=task()):
    params = deepcopy(params)

    vTh = params["neuron"]["v_thresh"]
    eL = params["neuron"]["v_rest"]
    Cm = params["neuron"]["cm"]
    tauE = params["syn_e"]["tau_syn"] if tau_syn_e is None else tau_syn_e
    tauI = params["syn_i"]["tau_syn"] if tau_syn_i is None else tau_syn_i

    params["neuron"]["g_leak"] = gL * Cm
    params["syn_e"]["w_syn"] = w_syn_e * Cm / tauE
    params["syn_e"]["e_rev_e"] = (vTh - eL) * e_rev_e + eL
    params["syn_i"]["w_syn"] = w_syn_i * Cm / tauI
    params["syn_i"]["e_rev_i"] = (vTh - eL) * e_rev_i + eL

    return params


class SimCondExp:
    def __init__(self):
        # Make sure the simulator is compiled
        curdir = os.path.dirname(os.path.abspath(__file__))
        with subprocess.Popen(["make", "-s", "-C", curdir,
                               "all"], stdout=subprocess.PIPE, stderr=subprocess.PIPE) as process:
            stdout, stderr = process.communicate()
            if process.wait() != 0:
                raise Exception(
                    "Error while compiling C++ simulator backend:\n" + stderr.decode('utf8')
                )

        # Initialize the persistent cache
        try:
            self.results = shelve.open("/tmp/.cache_sim_cond_exp_pipe.db")
        except:
            self.results = {}

        # Initialize internal caches
        self.tasks = {}

        # Start the subprocess
        executable = '/tmp/sim_cond_exp_pipe'
        self.process =  subprocess.Popen(
            [executable], stdin=subprocess.PIPE, stdout=subprocess.PIPE)
        self.open = True

        # Spawn a thread for writing to the process
        self.write_queue = []
        def write_thread_proc():
            while True:
                while len(self.write_queue) > 0:
                    self.process.stdin.write(self.write_queue.pop(0))
                    self.process.stdin.flush()
                if not self.open:
                    break
                time.sleep(10e-3)
        self.write_thread = threading.Thread(target=write_thread_proc)
        self.write_thread.start()

    def submit(self, task, callback):
        # Make sure the simulator subprocess is still alive
        if not self.open:
            raise Exception("Simulator instance is closed")

        # Do not modify the task
        task = deepcopy(task)

        # Discretize rate_e and rate_i to two decimal places, increasing the
        # number of cache hits
        task["input"]["rate_e"] = int(100 * task["input"]["rate_e"]) / 100
        task["input"]["rate_i"] = int(100 * task["input"]["rate_i"]) / 100

        # Serialise the task description and hash it
        task_str = json.dumps(task).encode('utf-8')
        h = hashlib.sha256(task_str).hexdigest()[0:16]

        # If the task was already executed just call the callback and return
        if h in self.results:
            callback(task, self.results[h])
            return

        # If a callback is already registered for this task, append it to the
        # list of callbacks
        if h in self.tasks:
            self.tasks[h]["callbacks"].append(callback)
            return

        # Register the callback for the given hash
        self.tasks[h] = {"task": task, "callbacks": [callback]}

        # Submit the task to the subprocess
        task["input"]["id"] = h
        task_str = json.dumps(task).encode('utf-8')
        self.write_queue.append(task_str + b'\n')

    def wait(self):
        if self.open:
            response = self.process.stdout.readline()
            if response == "":
                # Wait for the write thread to finish
                self.open = False
                self.write_thread.join()

                # Explicitly close stdin
                self.process.stdin.close()

                self.process.wait()
                return False
            self.handle_response(response)
            return True
        return False

    def handle_response(self, response):
        # Split the response at the comma
        if (len(response) == 0):
            return
        data = response.decode('utf8')[:-1].split(',')

        # Extract the task id
        h, result = data[0], np.array(list(map(float, data[1:])))
        if h in self.tasks:
            # Call all associated callbacks
            task = self.tasks[h]
            for callback in task["callbacks"]:
                callback(task["task"], result)

            # Store the result for future reference
            self.results[h] = result

            del self.tasks[h]

    def close(self):
        if self.open:
            # Mark this instance as closed
            self.open = False
            self.write_thread.join()

            # Explicitly close stdin
            self.process.stdin.close()

            # Wait for final responses and handle them
            for response in self.process.stdout:
                self.handle_response(response)

            # Wait for the process to exit and explicitly close it
            self.process.wait()

            # Close the database
            if hasattr(self.results, 'close'):
                self.results.close()

    def __enter__(self):
        return self

    def __exit__(self, type, value, traceback):
        self.close()

File: nengo_conductance_synapses/sim_cond_exp/sim_cond_exp_pipe.cpp
/*
 *  This file is part of soft_cond_lif
 *
 *  soft_cond_lif is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  soft_cond_lif is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with soft_cond_lif.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <atomic>
#include <iostream>
#include <mutex>
#include <queue>
#include <string>
#include <thread>
#include <vector>

#include <cinder/cinder.hpp>

#include "json.hpp"

using namespace cinder;
using json = nlohmann::json;

/**
 * Uses RTTI information to convert a JSON object to a Cinder typed vector\
 * structure.
 */
template <typename T>
T json_to_typed_vector(const json &j)
{
	constexpr auto names = T::names();
	T res;
	for (size_t i = 0; i < T::size(); i++) {
		if (j.find(names[i]) != j.end()) {
			res[i] = j[names[i]];
		}
	}
	return res;
}

/**
 * Structure holding all information for a single experiment run.
 */
struct Task {
	std::string id;
	Time T;
	Real rate_e;
	Real rate_i;
	size_t repeat;
	size_t seed;
	Real spike_loss_e;
	Real spike_loss_i;
	LIFParameters lif_params;
	CondExpParameters syn_e_params;
	CondExpParameters syn_i_params;
};

int main()
{
	std::queue<Task> queue;
	std::mutex queue_mutex, io_mutex;
	std::atomic<bool> cancel(false);

	auto f = [&]() {
		while (true) {
			Task current_task;

			// Fetch a task
			while (true) {
				{
					std::lock_guard<std::mutex> lock(queue_mutex);
					if (!queue.empty()) {
						current_task = queue.front();
						queue.pop();
						break;
					}
				}
				if (cancel) {
					return;  // No task available, we're done
				}
				// Wait a short while for another task
				using namespace std::chrono_literals;
				std::this_thread::sleep_for(10ms);
			}

			// Some shorthands
			const Real r_e = current_task.rate_e;
			const Real r_i = current_task.rate_i;
			const size_t seed = current_task.seed;
			const Time T = current_task.T;

			std::vector<Real> results;

			for (size_t j = 0; j < current_task.repeat; j++) {
				// Generate input spike trains
				auto spikes_e = simulate_spike_loss(
				    poisson_spike_train(0_s, T, std::abs(r_e),
				                        seed + 2 * j + 0),
				    current_task.spike_loss_e, seed + 2 * j + 0);
				auto spikes_i = simulate_spike_loss(
				    poisson_spike_train(0_s, T, std::abs(r_i),
				                        seed + 2 * j + 1),
				    current_task.spike_loss_i, seed + 2 * j + 1);

				// Negate the weight for negative rates
				CondExpParameters sep = current_task.syn_e_params;
				CondExpParameters sip = current_task.syn_i_params;
				sep[CondExpParameters::idx_w_syn] *= (r_e > 0.0) ? 1.0 : -1.0;
				sip[CondExpParameters::idx_w_syn] *= (r_i > 0.0) ? 1.0 : -1.0;

				EulerIntegrator integrator;
				NullRecorder recorder;
				NullController controller;
				auto current_source = make_current_source(
				    CondExp(sep, spikes_e), CondExp(sip, spikes_i));

				size_t spike_count = 0;
				auto neuron = make_neuron<LIF>(current_source,
				                               [&spike_count, T](Time t) {
					                               if (t > T * 0.5) {
						                               spike_count++;
					                               }
					                           },
				                               current_task.lif_params);

				make_solver(neuron, integrator, recorder, controller)
				    .solve(T, 0.1_ms);

				results.emplace_back(2.0 * Real(spike_count) / T.sec());
			}

			// Write the result to std::cout
			{
				std::lock_guard<std::mutex> lock(io_mutex);
				std::cout << current_task.id;
				for (Real res : results) {
					std::cout << "," << res;
				}
				std::cout << std::endl;
			}
		}
	};

	// Spawn n threads
	std::vector<std::thread> threads;
	for (size_t i = 0; i < std::thread::hardware_concurrency(); i++) {
		threads.emplace_back(std::thread(f));
	}

	std::string line;
	while (std::getline(std::cin, line)) {
		// Skip empty lines
		if (line.empty()) {
			continue;
		}

		// Parse the given line into a JSON datastructure
		json j = json::parse(line);

		// Read the task from the JSON object
		Task t;
		t.id = j["input"]["id"];
		t.T = Time::sec(j["input"]["T"]);
		t.rate_e = j["input"]["rate_e"];
		t.rate_i = j["input"]["rate_i"];
		t.repeat = j["input"]["repeat"];
		t.seed = j["input"]["seed"];
		t.spike_loss_e = j["input"]["spike_loss_e"];
		t.spike_loss_i = j["input"]["spike_loss_i"];
		t.lif_params = json_to_typed_vector<LIFParameters>(j["neuron"]);
		t.syn_e_params = json_to_typed_vector<CondExpParameters>(j["syn_e"]);
		t.syn_i_params = json_to_typed_vector<CondExpParameters>(j["syn_i"]);

		// Emplace the task on the task queue
		{
			std::lock_guard<std::mutex> lock(queue_mutex);
			queue.emplace(t);
		}
	}

	// Wait for all threads to finish
	cancel = true;
	for (size_t i = 0; i < threads.size(); i++) {
		threads[i].join();
	}

	return 0;
}

